docs/contributing: Initial Provider Design page (#17034)
hashicorp/terraform-provider-aws: docs/contributing/provider-design.md

Please note: the overall design and implementation across all AWS functionality is federated. This means that individual services may implement concepts and use terminology differently. As such, this guide may not be fully exhaustive in all situations. The goal is to provide enough hints about these concepts and towards specific terminology to point contributors in the correct direction, especially when researching prior implementations.

### Authorization and Acceptance Resources

> Certain AWS services use an authorization-acceptance model for cross-account associations or access. Some examples of these can be found with:

Reviewer: ```suggestion
Some AWS services use an authorization-acceptance model for cross-account associations or access. Examples include:
```

* Direct Connect Association Proposals
* GuardDuty Member Invitations
* RAM Resource Share Associations
* Route 53 VPC Associations
* Security Hub Member Invitations

> This type of API model either implements a form of an invitation/proposal identifier that can be used to accept the authorization, otherwise it may just be a matter of providing the desired AWS Account identifier as the target of an authorization. In the latter case, the acceptance is implicit when creating the other half of the association.

Reviewer: ```suggestion
Depending on the API and components, AWS uses two basic ways of creating cross-account associations. One way is for a practitioner to generate an invitation (i.e., proposal) identifier in an AWS account. A practitioner then uses the identifier to accept the invitation in a separate AWS account. The second way is for a practitioner, from one AWS account, to identify a different AWS account. The second way does not require an explicit acceptance to create the association.
```

> To model the separate invitation/proposal in Terraform resources:

Reviewer: ```suggestion
To model creating an association using an invitation or proposal, follow these guidelines.
```

> * Typically the authorization side has separate and sufficient creation and read API functionality to create an "invitation" or "proposal" resource.
> * If the acceptance side has separate and sufficient accept, read, and reject API functionality, then an "accepter" resource may be created, where the operations are mapped as:
>     * Create: Accepts the invitation/proposal.
>     * Read: Reads the invitation/proposal to determine status. If not found, then it should fallback to reading the API resource associated with the invitation/proposal. As evidenced in some previous API implementations, the invitation/proposal may be temporary and removed from the API after an indeterminate amount of time which may not be documented or easily discoverable from testing.
>     * Delete: Rejects or otherwise deletes the invitation/proposal.

Reviewer: ```suggestion
* Follow the naming in the AWS service API to determine whether to use the term "invitation" or "proposal."
* For the originating account, create an "invitation" or "proposal" resource. Make sure that the AWS service API has operations for creating and reading invitations. 
* For the responding account, create an "accepter" resource. Ensure that the API has operations for accepting, reading, and rejecting invitations in the responding account. Map the operations as follows:
	* Create: Accepts the invitation.
	* Read: Reads the invitation to determine its status. Note that in some APIs, invitations expire and disappear, complicating associations. If a resource does not find an invitation, the developer should implement a fall back to read the API resource associated with the invitation/proposal.
	* Delete: Rejects or otherwise deletes the invitation.
```

> Otherwise to model an implicit acceptance in Terraform resources, usually there is an "association" resource with an optional "authorization" resource, each mapped to the create/read/delete functionality of the API for the associated part of the relationship.

Reviewer: ```suggestion
To model the second type of association, implicit associations, create an "association" resource and, optionally, an "authorization" resource. Map create, read, and delete to the corresponding operations in the AWS service API.
```

### Cross-Service Functionality

> Many AWS service APIs build on top of other AWS services. Some examples of these include:
> 
> * EKS Node Groups managing Auto Scaling Groups
> * Lambda Functions managing EC2 ENIs
> * Transfer Servers managing EC2 VPC Endpoints
> 
> In certain of these cross-service API implementations, there can be lacking management or description capabilities of the other service functionality, which can make the Terraform resource implementation seem incomplete or unsuccessful in end-to-end configurations. Given the overall “resources should represent a single API object” goal from the [HashiCorp Provider Design Principles](https://www.terraform.io/docs/extend/hashicorp-provider-design-principles.html), it is expected that a single resource in this provider will only communicate with a single AWS service API in a self-contained manner. As such, cross-service resources will not be implementated.
> 
> In particular, some of the rationale behind this design decision is:
> 
> * Unexpected IAM permissions being necessary for the resource. In restrictive environments, these permissions may not be available or acceptable for security purposes.
> * Unexpected CloudTrail logs being generated for the resource.
> * Unexpected API endpoints configuration being necessary for those using custom endpoints, such as VPC endpoints.
> * Unexpected changes to the AWS service internals. Given that these cross-service implementations require being based on internal details the current AWS service implementation, these details can change over time and may not be considered as a breaking change by the AWS service API during an API upgrade.
> 
> As a poignant real world example of the last point, the provider was relying on the description of Lambda Function created EC2 ENIs for the purposes of trying to delete lingering ENIs due to a common misconfiguration. This functionality was helpful for operators as the issue was hard to diagnose due to a generic EC2 API error message. Years after the implementation, the Lambda API was updated and immediately many operators started reporting Terraform executions were failing. Downgrading the provider could not help, many configurations depended on resources and behaviors in recent versions. For other environments that were running many versions behind, forcing an immediate provider upgrade with the fix could also have other unrelated or unexpected changes. In the end, both HashiCorp and AWS needed to perform a large scale outreach about upgrading or fixing the misconfiguration, causing lost time for many operators and the provider maintainers.

Author: Mostly the same, just the last bullet point. I think stressing the "internals" is important.

```suggestion
Many AWS service APIs build on top of other AWS services. Some examples of these include:

* EKS Node Groups managing Auto Scaling Groups
* Lambda Functions managing EC2 ENIs
* Transfer Servers managing EC2 VPC Endpoints

Some cross-service API implementations lack the management or description capabilities of the other service. The lack can make the Terraform resource implementation seem incomplete or unsuccessful in end-to-end configurations. Given the overall “resources should represent a single API object” goal from the [HashiCorp Provider Design Principles](https://www.terraform.io/docs/extend/hashicorp-provider-design-principles.html), a resource must only communicate with a single AWS service API. As such, maintainers will not approve cross-service resources.

The rationale behind this design decision includes the following:

* Unexpected IAM permissions being necessary for the resource. In high-security environments, all the service permissions may not be available or acceptable.
* Unexpected services generating CloudTrail logs for the resource.
* Needing extra and unexpected API endpoints configuration for organizations using custom endpoints, such as VPC endpoints.
* Unexpected changes to the AWS service internals for the cross-service implementations. Given that this functionality is not part of the primary service API, these details can change over time and may not be considered as a breaking change by the service team for an API upgrade.

A poignant real-world example of the last point involved a Lambda resource. The resource helped clean up extra resources (ENIs) due to a common misconfiguration. Practitioners found the functionality helpful since the issue was hard to diagnose. Years later, AWS updated the Lambda API. Immediately, practitioners reported that Terraform executions were failing. Downgrading the provider was not possible since many configurations depended on recent releases. For environments running many versions behind, forcing an upgrade with the fix would likely cause unrelated and unexpected changes. In the end, HashiCorp and AWS performed a large-scale outreach to help upgrade and fixing the misconfigurations. Provider maintainers and practitioners lost considerable time.
```

### IAM Resource-Based Policy Resources

> Many AWS resources support the concept of an [IAM resource-based policy](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_identity-vs-resource.html) for specifying IAM permissions associated the resource. Some examples of these include:
> 
> * ECR Repository Policies
> * EFS File System Policies
> * SNS Topic Policies
> 
> When implementing this support in the Terraform AWS Provider, it is preferred to create a separate resource for a few reasons:
> 
> * Many of these policies require the Amazon Resource Name (ARN) of the resource in the policy itself. It is difficult to workaround this requirement with custom difference handling within a self-contained resource.
> * Sometimes policies between two resources need to be written where they cross-reference each other resource's ARN within each policy. Without a separate resource, this introduces a configuration cycle.
> * Splitting the resources allows operators to logically split their infrastructure on purely operational and security boundaries with separate configurations/modules.
> * Splitting the resources prevents any separate policy API calls from needing to be permitted in the main resource in environments with restrictive IAM permissions, which can be undesirable.
> 
> In rare cases, it may be necessary to implement the policy handling within the associated resource itself. This should only be reserved for cases where the policy is _required_ during resource creation.

Author: Following the earlier suggestion of "restrictive" to "high-security" (removing the qualifier altogether)

```suggestion
For some AWS components, the AWS API allows specifying an [IAM resource-based policy](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_identity-vs-resource.html), the IAM policy to associate with a component. Some examples include:

* ECR Repository Policies
* EFS File System Policies
* SNS Topic Policies

Provider developers should implement this capability in a new resource rather than adding it to the associated resource. Reasons for this include:

* Many of the policies must include the Amazon Resource Name (ARN) of the resource. Working around this requirement with custom difference handling within a self-contained resource is unnecessarily cumbersome.
* Some policies involving multiple resources need to cross-reference each other's ARNs. Without a separate resource, this introduces a configuration cycle.
* Splitting the resources allows operators to logically split their configurations into purely operational and security boundaries. This allows environments to have distinct practitioners roles and permissions for IAM versus infrastructure changes.

One rare exception to this guideline is where the policy is _required_ during resource creation.
```

### Task Execution and Waiter Resources

Certain AWS functionality operates with the execution of individual tasks or external validation processes that have a completion status. Some examples of these include:

* ACM Certificate validation
* EC2 AMI copying
* RDS DB Cluster Snapshot management

The general expectation in these cases is that the AWS service APIs provide sufficient lifecycle management to start the execution/validation and read the status of that operation. At the end of the operation, the result may be something that can be fully managed on its own with update and delete support (e.g. with EC2 AMI copying, a copy of the AMI is a manageable AMI). In these cases, the preference is to create a separate Terraform resource, rather than implement the functionality within the parent Terraform resource. This recommendation is even at the expense of duplicated resource and schema handling.

One related consideration is the ability to manage the running state of a resource (e.g. start/stop it). See the [Managing Resource Running State section](#managing-resource-running-state) for specific details.

## Other Considerations

### AWS Credential Exfiltration

> In the security-minded interest of the community, the provider will not implement data sources that will give operators the ability to reference or export the AWS credentials of the running provider. While there are valid use cases of this information, such as using those AWS credentials to execute AWS CLI calls as part of the same Terraform configuration, it also becomes possible for those credentials to be discovered and used completely outside of the current Terraform execution. Some of the concerns around this topic include:
> 
> * The values can potentially be visible in Terraform user interface output or logging where anyone with access to those can see and use them.
> * The values must be stored in the Terraform state in plaintext, due to how Terraform operates. Anyone with access or another Terraform configuration that references the state can see and use them.
> * Any new functionality related to this, while opt-in to implement, is also opt-in to prevent via security controls or policies. Since this introduces a weak by default security posture, organizations wishing to implement those controls may require advance notice before the release or may never be able to update.

Author: Minor tweaks

```suggestion
In the interest of security, the maintainers will not approve data sources that provide the ability to reference or export the AWS credentials of the running provider. There are valid use cases for this information, such as to execute AWS CLI calls as part of the same Terraform configuration. However, this mechanism may allow credentials to be discovered and used outside of Terraform. Some specific concerns include:

* The values may be visible in Terraform user interface output or logging, allowing anyone with user interface or log access to see the credentials.
* The values are currently stored in plaintext in the Terraform state, allowing anyone with access to the state file or another Terraform configuration that references the state access to the credentials.
* Any new related functionality, while opt-in to implement, is also opt-in to prevent via security controls or policies. Adopting a weaker default security posture requires advance notice and prevents organizations that implement those controls from updating to a version with any such functionality.
```

### Authorization and Acceptance Resources

Some AWS services use an authorization-acceptance model for cross-account associations or access. Examples include:

* Direct Connect Association Proposals
* GuardDuty Member Invitations
* RAM Resource Share Associations
* Route 53 VPC Associations
* Security Hub Member Invitations

Depending on the API and components, AWS uses two basic ways of creating cross-region and cross-account associations. One way is to generate an invitation (or proposal) identifier from one AWS account to another. Then in the other AWS account, that identifier is used to accept the invitation. The second way is configuring a reference to another AWS account identifier. These may not require explicit acceptance on the receiving account to finish creating the association or begin working.

To model creating an association using an invitation or proposal, follow these guidelines.

* Follow the naming in the AWS service API to determine whether to use the term "invitation" or "proposal."

> * For the originating account, create an "invitation" or "proposal" resource. Make sure that the AWS service API has operations for creating and reading invitations. 

Author: ```suggestion
* For the originating account, create an "invitation" or "proposal" resource. Make sure that the AWS service API has operations for creating and reading invitations.
```

* For the responding account, create an "accepter" resource. Ensure that the API has operations for accepting, reading, and rejecting invitations in the responding account. Map the operations as follows:

> 	* Create: Accepts the invitation.
> 	* Read: Reads the invitation to determine its status. Note that in some APIs, invitations expire and disappear, complicating associations. If a resource does not find an invitation, the developer should implement a fall back to read the API resource associated with the invitation/proposal.
> 	* Delete: Rejects or otherwise deletes the invitation.

Author: ```suggestion
    * Create: Accepts the invitation.
    * Read: Reads the invitation to determine its status. Note that in some APIs, invitations expire and disappear, complicating associations. If a resource does not find an invitation, the developer should implement a fall back to read the API resource associated with the invitation/proposal.
    * Delete: Rejects or otherwise deletes the invitation.
```

One rare exception to this guideline is where the policy is _required_ during resource creation.

### Managing Resource Running State

The AWS API provides the ability to start, stop, enable, or disable some AWS components. Some examples include:

* Batch Job Queues
* CloudFront Distributions
* RDS DB Event Subscriptions

In this situation, provider developers should implement this ability within the resource instead of creating a separate resource. Since a practitioner cannot practically manage interaction with a resource's states in Terraform's declarative configuration, developers should implement the state management in the resource. This design provides consistency and future-proofing even where updating a resource in the current API is not problematic.

### Task Execution and Waiter Resources

Some AWS operations are asynchronous. Terraform requests that AWS perform a task. Initially, AWS only notifies Terraform that it received the request. Terraform then requests the status while awaiting completion. Examples of this include:

* ACM Certificate validation
* EC2 AMI copying
* RDS DB Cluster Snapshot management

> In this situation, provider developers should create a separate resource representing the task, assuming that the AWS service API provides operations to start the task and read its status. Adding the task functionality to the parent resource muddies its infrastructure-management purpose. The maintainers prefer this approach even though there is some duplication of an existing resource. For example, the provider has a resource for copying an EC2 AMI in addition to the EC2 AMI resource itself. This modularity allows practitioners to manage the result of the task resource with another resource. 

Author: ```suggestion
In this situation, provider developers should create a separate resource representing the task, assuming that the AWS service API provides operations to start the task and read its status. Adding the task functionality to the parent resource muddies its infrastructure-management purpose. The maintainers prefer this approach even though there is some duplication of an existing resource. For example, the provider has a resource for copying an EC2 AMI in addition to the EC2 AMI resource itself. This modularity allows practitioners to manage the result of the task resource with another resource.
```

* Functionality that requires additional software beyond the Terraform AWS Provider to be installed on the host executing Terraform. This currently includes the AWS CLI. See the [Terraform External Provider](https://registry.terraform.io/providers/hashicorp/external/latest) and other available providers instead.

## Infrastructure as Code Suitability

While a goal of the provider is to gain as much resource coverage as possible, from a design perspective, it is necessary to be pragmatic that there are APIs or portions of APIs that may not be wholly compatible with infrastructure as code concepts and managing them with that methodology. In these cases, it is recommended to open an AWS Support case (potentially encouraging others to do the same) to either fix or enhance the AWS service API so components are more self-contained and compatible with these concepts. Sometimes these types of AWS Support cases can also provide some good insight into the AWS service expectations or API behaviors that may not be well documented.

## Resource Type Considerations

> The general heuristic of when to create or not create a separate Terraform resource is if the AWS service API implements create, read, and delete operations for a component. Terraform resources work best as the smallest blocks of infrastructure, where configurations can build abstractions on top of these such as [Terraform Modules](https://www.terraform.io/docs/modules/). However, not all AWS service API functionality falls cleanly into components to be managed with those types of operations. It may be necessary to consider other patterns of mapping API operations to Terraform resources and operations. This section highlights these design recommendations, such as when to consider implemantations within the same Terraform resource or as a separate Terraform resource.

Reviewer 2: ```suggestion
The general heuristic of whether or not to create a separate Terraform resource is if the AWS service API implements create, read, and delete operations for a component. Terraform resources work best as the smallest blocks of infrastructure, where configurations can build abstractions on top of these such as [Terraform Modules](https://www.terraform.io/docs/modules/). However, not all AWS service API functionality falls cleanly into components to be managed with those types of operations. It may be necessary to consider other patterns of mapping API operations to Terraform resources and operations. This section highlights these design recommendations, such as when to consider implemantations within the same Terraform resource or as a separate Terraform resource.
```

### Authorization and Acceptance Resources

Certain AWS services use an authorization-acceptance model for cross-account associations or access. Some examples of these can be found with:

* Direct Connect Association Proposals
* GuardDuty Member Invitations
* RAM Resource Share Associations
* Route 53 VPC Associations
* Security Hub Member Invitations

This type of API model either implements a form of an invitation/proposal identifier that can be used to accept the authorization, otherwise it may just be a matter of providing the desired AWS Account identifier as the target of an authorization. In the latter case, the acceptance is implicit when creating the other half of the association.

To model the separate invitation/proposal in Terraform resources:

> * Typically the authorization side has separate and sufficient creation and read API functionality to create an "invite" or "proposal" resource.

Reviewer 2: Keep consistent.

```suggestion
* Typically the authorization side has separate and sufficient creation and read API functionality to create an "invitation" or "proposal" resource.
```

* If the acceptance side has separate and sufficient accept, read, and reject API functionality, then an "accepter" resource may be created, where the operations are mapped as:
    * Create: Accepts the invitation/proposal.

>     * Read: Reads the invitation/proposal to determine status. If not found, then it should fallback to reading the API resource associated with the invitation/proposal/association. As evidenced in some previous API implementations, the invitation/proposal may be temporary and removed from the API after an indeterminate amount of time which may not be documented or easily discoverable from testing.

Reviewer 2: As `association` has not yet been mentioned, should `invitation/proposal/association` be `invitation/proposal`?

Author: ```suggestion
    * Read: Reads the invitation/proposal to determine status. If not found, then it should fallback to reading the API resource associated with the invitation/proposal. As evidenced in some previous API implementations, the invitation/proposal may be temporary and removed from the API after an indeterminate amount of time which may not be documented or easily discoverable from testing.
```

    * Delete: Rejects or otherwise deletes the invitation/proposal.

Otherwise, to model an implicit acceptance in Terraform resources:

* If the authorization side has separate and sufficient creation and read API functionality, an "authorization" resource is created.

> * An "association" resource is created, where the operations are mapped as:

Reviewer 2: Are these 2 sentences trying to say the same thing?

Author: I'll combine these into the above "Otherwise" sentence -- they are different, but my writing was unclear.

Reviewer 2: IMHO it's more understandable now that the relationship between association and authorization is spelled out.
